libr/anal/type.c

```diff
@@ -0,0 +1,110 @@
+/* radare - LGPL - Copyright 2019 - pancake, oddcoder, Anton Kochkov */
+
+#include <string.h>
+#include <r_anal.h>
+#include "sdb/sdb.h"
+
+/*!
+ * \brief Save the size of the given datatype in sdb
+ * \param sdb_types pointer to the sdb for types
+ * \param name the datatype whose size if to be stored
+ */
+static void save_type_size(Sdb *sdb_types, char *name) {
+	char *type = NULL;
+	r_return_if_fail (sdb_types && name);
+	if (!sdb_exists (sdb_types, name) || !(type = sdb_get (sdb_types, name, 0))) {
+		return;
+	}
+	char *type_name_size = r_str_newf ("%s.%s.%s", type, name, "!size");
+	r_return_if_fail (type_name_size);
+	int size = r_type_get_bitsize (sdb_types, name);
+	sdb_set (sdb_types, type_name_size, sdb_fmt ("%d", size), 0);
+	free (type);
+	free (type_name_size);
+}
+
+/*!
+ * \brief Save the sizes of the datatypes which have been parsed
+ * \param core pointer to radare2 core
+ * \param parsed the parsed c string in sdb format
+ */
+static void save_parsed_type_size(RAnal *anal, const char *parsed) {
+	r_return_if_fail (anal && parsed);
+	char *str = strdup (parsed);
+	if (str) {
+		char *ptr = NULL;
+		int offset = 0;
+		while ((ptr = strstr (str + offset, "=struct\n")) ||
+				(ptr = strstr (str + offset, "=union\n"))) {
+			*ptr = 0;
+			if (str + offset == ptr) {
+				break;
+			}
+			char *name = ptr - 1;
+			while (name > str && *name != '\n') {
+				name--;
+			}
+			if (*name == '\n') {
+				name++;
+			}
+			save_type_size (anal->sdb_types, name);
+			*ptr = '=';
+			offset = ptr + 1 - str;
+		}
+		free (str);
+	}
+}
+
+R_API void r_anal_remove_parsed_type(RAnal *anal, const char *name) {
+	r_return_if_fail (anal && name);
+	Sdb *TDB = anal->sdb_types;
+	SdbKv *kv;
+	SdbListIter *iter;
+	const char *type = sdb_const_get (TDB, name, 0);
+	if (!type) {
+		return;
+	}
+	int tmp_len = strlen (name) + strlen (type);
+	char *tmp = malloc (tmp_len + 1);
+	r_type_del (TDB, name);
+	if (tmp) {
+		snprintf (tmp, tmp_len + 1, "%s.%s.", type, name);
+		SdbList *l = sdb_foreach_list (TDB, true);
+		ls_foreach (l, iter, kv) {
+			if (!strncmp (sdbkv_key (kv), tmp, tmp_len)) {
+				r_type_del (TDB, sdbkv_key (kv));
+			}
+		}
+		ls_free (l);
+		free (tmp);
+	}
+}
+
+R_API void r_anal_save_parsed_type(RAnal *anal, const char *parsed) {
+	r_return_if_fail (anal && parsed);
+	// First, if this exists, let's remove it.
+	char *type = strdup (parsed);
+	if (type) {
+		char *name = NULL;
+		if ((name = strstr (type, "=type")) ||
```

Comment:
Move into separate function. This seems omething we can simplify if we prefix al the tkeywords with $ or so